config/src/test/java/org/springframework/security/config/annotation/web/configurers/AnonymousConfigurerTests.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.config.annotation.web.configurers;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+import org.springframework.security.config.test.SpringTestRule;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Controller;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.MvcResult;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseBody;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+
+/**
+ * @author Rob Winch
+ * @author Josh Cummings
+ */
+public class AnonymousConfigurerTests {
+	@Rule
+	public final SpringTestRule spring = new SpringTestRule();
+
+	@Autowired
+	private MockMvc mockMvc;
+
+	@Test
+	public void requestWhenAnonymousTwiceInvokedThenDoesNotOverride() throws Exception {
+		this.spring.register(InvokeTwiceDoesNotOverride.class).autowire();
+
+		MvcResult result = this.mockMvc.perform(get("/")).andReturn();
+
+		assertThat(result.getResponse().getContentAsString()).isEqualTo("custom");
+	}
+
+	@EnableWebSecurity
+	static class InvokeTwiceDoesNotOverride extends WebSecurityConfigurerAdapter {
+
+		@Override
+		protected void configure(HttpSecurity http) throws Exception {
+			http
+				.anonymous()
+					.principal("custom")
+					.and()
+				.anonymous();
+		}
+
+		@Controller
+		static class PrincipalController {
+			@RequestMapping("/")
+			@ResponseBody String principal() {
+				Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
```

Comment:
Make `Authentication` an argument to the method rather than resolving through `SecurityContextHolder` which requires a static method to work

Reply:
I originally attempted this, as I feel the same way; however `Authentication` doesn't appear to resolve, e.g. the method param is always `null`. I also attempted with `@AuthenticationPrincipal String principal` as well as simply `SecurityContext`. Happy to dig into this further, but do you happen to see another configuration piece I am missing?

Reply:
Looks like if I add `@EnableWebMvc` to my configuration, then `AuthenticationPrincipalArgumentResolver` gets wired, and now I can use `@AuthenticationPrincipal`, which does clean up the code nicely. Does that sound as you would expect?